src/coreclr/jit/optimizer.cpp

```diff
@@ -5512,6 +5512,14 @@ void Compiler::optPerformHoistExpr(GenTree* origExpr, unsigned lnum)
     compCurBB = preHead;
     hoist     = fgMorphTree(hoist);
 
+#ifdef DEBUG
+    if ((JitConfig.JitDoEarlyProp() == 0) || opts.optRepeat)
+    {
+        // Normally its EarlyProp's responsibility to propagate these trees/flags
+        preHead->bbFlags |= BBF_HAS_NULLCHECK | BBF_HAS_IDX_LEN;
```

Comment:
The comment is incorrect. EarlyProp doesn't propagate these flags. It uses them and also asserts if they are not set.

Reply:
The condition is also incorrect. If `(JitConfig.JitDoEarlyProp() == 0)` we don't need to maintain these flags since early prop is the only phase that uses them.

I would just maintain these flags properly. Propagate the flags both in debug and release if the the basic block of `origExpr` has them.

Reply:
What is the advantage in maintaining state that will not be used (I realize we could start using them later in the phase order, but then we'd have to write a proper checker anyway)?

FWIW, my take on the fix is to simply reset the flags in `ResetOptAnnotations` (via a visitor).

Reply:
Right, I would just always propagate these like @erozen says.

It is consistent with what we do elsewhere.

Reply:
> What is the advantage in maintaining state that will not be used (I realize we could start using them later in the phase order, but then we'd have to write a proper checker anyway)?

Under optRepeat this state will be used by early prop on the next iteration.

Reply:
Yes, of course, the bug still has to be fixed after all.

To put my question another way: how do we decide what we should maintain for `optRepeat` (as a non-shipping configuration) and what we should just forget/reset/recompute?

Reply:
If it is cheap and easy to do the least surprising thing, we should do it.

Reply:
`optRepeat` is non-shipping but we do experiments with it to assess the usefulness of phases and possible phase reorderings. In this case not maintaining these flags may lead to some missed optimization opportunities and misleading results from optRepeat experiments. And, as @AndyAyersMS said, it's easy to maintain the flags in this case.